### PR TITLE
Include DOI in search URIs as a "doi:" URI

### PR DESCRIPTION
**Pull request by judell**
Addresses the problem, felt most acutely by @memartone and her SciCrunch colleagues (hi, @tgbugs), of the client not finding SciBot annotations on variants and syndicated copies of articles.

I think it may be redundant to include the non-DOI URI in the list of URIs to search, and originally I didn't, but I don't think it's harmful, and including both more closely matches how we search for PDFs (URI plus fingerprint).

### CODE REVIEW

**Comment by robertknight** [comment]
Thanks @judell. When working on the client or service, you may find it helpful to have ESLint (JavaScript) & flake8 (Python) integrations installed for your editor. This gives you live feedback on common errors and coding style issues. Atom and Visual Studio Code in particular have very easy one-click installs for these.

**Comment by robertknight** [comment]
_These are some notes I made while taking a first look at this:_

----

For the purpose of testing this, I created a [metabase card](https://metabase.hypothes.is/card/132) which lists the HTTP URLs and IDs of public annotations which are associated with a DOI.

For example, this article http://aac.asm.org/content/49/2/479.full contains the following `<meta>` tags in the HTML:

```html
<meta content="10.1128/AAC.49.2.479-487.2005" name="DC.Identifier" />
<meta content="10.1128/AAC.49.2.479-487.2005" name="citation_doi" />
```

These entries are extracted from the page by [document.coffee](https://github.com/hypothesis/client/blob/0454367dfb9abca8c6f84d0ddc1a3faf8b8d49b9/src/annotator/plugin/document.coffee). In particular I think it is [this code in getLinks()](https://github.com/hypothesis/client/blob/0454367dfb9abca8c6f84d0ddc1a3faf8b8d49b9/src/annotator/plugin/document.coffee#L147) which extracts DOIs and adds a "doi:" prefix.

When the annotation is POSTed to the server, entries are created in the `document_uri` table which maps the annotation's main ("target") URI to the "doi:" URL. eg. for this article the _claimant_ is `http://aac.asm.org/content/49/2/479.full ` and the URI is `doi:10.1128/AAC.49.2.479-487.2005`.

Given a DOI URI you can retrieve the associated annotations via a request to the search API. eg: https://hypothes.is/api/search?uri=doi:10.1128/AAC.49.2.479-487.2005

Previously the "doi:" URI was not being submitted to the server when the client searched for annotations. With this change here, the client will include "doi:" URLs in the above search query whenever it finds appropriate `<meta>` tags on the page. This enables all annotations that relate to content labeled with a specific DOI to be shown when viewing a page that identifies itself as being associated with that DOI.

The same approach could work with other types of identifier such as ISBNs (books), arXiv IDs (preprints on arXiv.org) and more.

**Comment by judell** [comment]
> The same approach could work with other types of identifier such as ISBNs (books), arXiv IDs 
> (preprints on arXiv.org) and more.

Yes. In particular, we've having an ongoing conversation with OER publishers after an initial session at the Hewlett-sponsored summit. They want annotations on textbooks syndicated across their network to coalesce. DOI seems not to be the solution in their case because they'd need to mint them for chapters, or even sections, which would be overkill. rel="canonical" is likewise not the solution because publisher B won't want to declare a publisher A URL canonical.

So they are considering some way to reflect the canonical URL for a page into other metadata, perhaps dc:identifier. That's still to be decided but to your point, whatever is agreed on becomes a thing we could look up in this same way.

**Comment by judell** [comment]
>  ESLint

Got it, thx. I see that not only does https://www.npmjs.com/package/eslint-config-hypothesis exist, but it seems rather popular!

**Comment by judell** [comment]
Thanks so much for these notes! I've been wanting to make a start on documenting all this but it has felt overwhelming. Now that you've broken the ice I'll try to jump in.

We honor two flavors of DOI declarations which we know, in the database, as `document_uri.type` of `highwire-doi` and `dc-doi`, and which appear in HTML as:

`<meta name="citation_doi" content="10.1234/journal123">`

and

`<meta name="DC.identifier"  content="http://loc.gov/ark:/12025/654xz321" />`

(Highwire Press defines citation_doi, as part of a namespace [widely adopted](https://scholar.google.com/intl/en/scholar/inclusion.html) in the scholarly world. DC.identifier is from [Dublin Core](http://www.ukoln.ac.uk/metadata/dcmi-ieee/identifiers/). 

Here's data on the distribution of both in our system: https://docs.google.com/spreadsheets/d/1mY5yn25T9zhj68CJqLAXlN_YYbkldsmp6zl7ZeET52E (this was prepared for Crossref who are looking to ingest DOI-related annotations).

Related: https://github.com/hypothesis/product-backlog/issues/144

File: src/sidebar/frame-sync.js [hunk 1]
```diff
@@ -174,6 +174,16 @@ function FrameSync($rootScope, $window, Discovery, annotationUI, bridge) {
         });
       }
 
+      // if doi available as highwire or dc meta, include in search
+
+      if ( info.metadata && info.metadata.highwire && info.metadata.highwire.doi ) {
+        searchUris = searchUris.concat('doi:' + info.metadata.highwire.doi[0]);
```
Comment (sean-roberts) [inline src/sidebar/frame-sync.js]: the only feedback I would have, though it's super small, is using `searchUris.push(val)` is preferred to `searchUris = searchUris.concat(val)` in this case. Less moving parts to reason about

Comment (judell) [inline src/sidebar/frame-sync.js]: Yup, that's better, thx.

File: src/sidebar/frame-sync.js [hunk 2]
```diff
@@ -174,6 +174,16 @@ function FrameSync($rootScope, $window, Discovery, annotationUI, bridge) {
...
       }
 
+      // if doi available as highwire or dc meta, include in search
+
+      if ( info.metadata && info.metadata.highwire && info.metadata.highwire.doi ) {
+        searchUris.push('doi:' + info.metadata.highwire.doi[0]);
+      }
+
+      if ( info.metadata && info.metadata.dc && info.metadata.dc.doi)  {
+        searchUris.push('doi:' + info.metadata.dc.doi[0]);
```
Comment (robertknight) [inline src/sidebar/frame-sync.js]: I note that there is already some code in document.coffee which populates the `info.metadata.link` Array with `doi:` URLs. We should only do DOI metadata => "doi:" URI extraction in one place, so we'll need to figure out whether that should happen here or in document.coffee.

**Comment by judell** [comment]
Closing in favor of a new PR.

**Pull request closed without merging**